### PR TITLE
Fix modes in or-patterns

### PR DESCRIPTION
**Pull request by lpw25**
Or-patterns were not using the correct modes, allowing local values to leak.

(This is on top of #48)

### CODE REVIEW

File: testsuite/tests/typing-local/local.ml [hunk 1]
```diff
@@ -1718,3 +1718,38 @@ let promote (local_ x) = +x
...
+|}]
+
+let (Some z, _, _) | (None, Some z, _)
+    | (None, None, z) = (Some (ref 0), (local_ (Some (ref 0))), (ref 0))
+[%%expect{|
+Line 1, characters 10-11:
+1 | let (Some z, _, _) | (None, Some z, _)
+              ^
+Error: This value escapes its region
+|}]
```
Comment (stedolan) [inline testsuite/tests/typing-local/local.ml:1755]: These two tests are identical. (Also, the error message is a bit unfortunate, it's pointing to the wrong z)

**Review by stedolan — APPROVED** [review]
lgtm, but needs a rebase.

**Comment by stedolan** [comment]
Thinking about the error message made me think of an alternative implementation. (I'm not insisting we do it this way, but maybe worth considering).

We could have `let_bound_idents_full` with the following signature:
```ocaml
 val let_bound_idents_full:
    value_binding list
    -> (Ident.t * Types.type_expr * (string loc * Types.Value_mode.t)) list
```
That is, instead of computing and storing the joined mode in the Typedtree, we keep the modes separate and return them all. The caller of `let_bound_idents_full` can check mode constraints separately, so that we can report the location of the particular occurrence of an ident that causes an issue, instead of reporting the first occurrence. Also, if the joined modes are not stored, we don't need to change the Typedtree, as we can traverse down to the Tpat_idents as needed.

**Comment by lpw25** [comment]
I've rewritten the PR to follow your suggestion since its much less invasive and gives a better error message.

**Pull request merged**